python/GafferArnoldUI/ArnoldShaderUI.py

```diff
@@ -350,16 +350,35 @@ def __plugMetadata( plug, name ) :
 	parameterPlugComponentKeys = set()
 
 	for name, metadata in __metadata.items() :
-		keys = ( nodeKeys, parametersPlugKeys, parameterPlugKeys, parameterPlugComponentKeys )[name.count( "." )]
+		level = name.count( "." )
+		keys = ( nodeKeys, parametersPlugKeys, parameterPlugKeys, parameterPlugComponentKeys )[level]
 		keys.update( metadata.keys() )
 
+		# Register user defaults using the syntax established by GafferSceneUI.ShaderUI
+		if not level == 2 :
+			continue
+
+		if not nodeType == GafferArnold.ArnoldShader :
+			continue
+
+		userDefault = metadata.get( "userDefault" )
+		if userDefault == None :
+			continue
+
+		nodeName, _, plugName = name.split( "." )
+		Gaffer.Metadata.registerValue( "ai:surface:%s:%s" % ( nodeName, plugName ), "userDefault", userDefault )
+
 	for key in nodeKeys :
 		Gaffer.Metadata.registerValue( nodeType, key, functools.partial( __nodeMetadata, name = key ) )
 
 	for key in parametersPlugKeys :
 		Gaffer.Metadata.registerValue( nodeType, "parameters", key, functools.partial( __plugMetadata, name = key ) )
 
 	for key in parameterPlugKeys :
+
+		if key == 'userDefault' :
```

Comment:
Could this, and all the other conditionals above, be an indication that it'd be better to not put "userDefault" into `__metadata` in the first place? Perhaps it makes more sense to call `Gaffer.Metadata.registerValue` directly at line 286?

Reply:
Updated. Do you like this better?

I guess the problem with this version is that it'll register the userDefaults on lights in the same way and lights don't support that syntax we're using here?

Reply:
And we hardcode the "surface" bit in the metadata - so it wouldn't work on lights anyway..?

Reply:
> Updated. Do you like this better?

Thanks - yes, this is much better. Everything dealing with user defaults is in one place, and we're not putting something in a data structure that then needs to be ignored by the main consumer of that structure.

> I guess the problem with this version is that it'll register the userDefaults on lights in the same way and lights don't support that syntax we're using here?

Seems pretty reasonable that we'd want to support it for lights too? Any reason we couldn't?

Reply:
> Seems pretty reasonable that we'd want to support it for lights too? Any reason we couldn't?

No, no reason why we couldn't support it. I was just noting that currently we don't, but the code I've added kinda suggests that we do. If you're cool with what I have now I'm happy, too :)